scipy/interpolate/_cubic.py

```diff
@@ -491,34 +570,177 @@ def __init__(self, x, y, axis=0, extrapolate=True):
-            A[-1, -2] = x[-1] - x[-3]
-            A[-1:, :-2] = dx[1:]
-
-            d = x[2] - x[0]
-            b[0] = ((dxr[0] + 2*d)*dxr[1]*slope[0] + dxr[0]**2*slope[1]) / d
+            # Filling the system for i=1..n-2
+            #                         (x[i-1] - x[i]) * s[i-1] +\
+            # 2 * ((x[i] - x[i-1]) + (x[i+1] - x[i])) * s[i]   +\
+            #                         (x[i] - x[i-1]) * s[i+1] =\
+            #       3 * ((x[i+1] - x[i])*(y[i] - y[i-1])/(x[i] - x[i-1]) +\
+            #           (x[i] - x[i-1])*(y[i+1] - y[i])/(x[i+1] - x[i]))
+
+            A[1, 1:-1] = 2 * (dx[:-1] + dx[1:])  # The diagonal
+            A[0, 2:] = dx[:-1]                   # The upper diagonal
+            A[-1, :-2] = dx[1:]                  # The lower diagonal
+
             b[1:-1] = 3 * (dxr[1:] * slope[:-1] + dxr[:-1] * slope[1:])
-            d = x[-1] - x[-3]
-            b[-1] = ((dxr[-1]**2*slope[-2] +
-                      (2*d + dxr[-1])*dxr[-2]*slope[-1]) / d)
 
-            s = solve_banded((1, 1), A, b, overwrite_ab=True, overwrite_b=True,
-                             check_finite=False)
+            bc_start, bc_end = bc
+
+            if bc_start == 'periodic':
+                # Due to the periodicity, and because y[-1] = y[0], the linear
+                # system has (n-1) unknowns/equations instead of n:
+                A = A[:, 0:-1]
+                A[1, 0] = 2 * (dx[-1] + dx[0])
+                A[0, 1] = dx[-1]
+
+                b = b[:-1]
+
+                # Also, due to the periodicity, the system is not tri-diagonal.
+                # We need to compute a "condensed" matrix of shape (n-2, n-2).
+                # See http://www.cfm.brown.edu/people/gk/chap6/node14.html for
+                # more explanations.
+                # The condensed matrix is obtained by removing the last column
+                # and last row of the (n-1, n-1) system matrix. The removed
+                # values are saved in scalar variables with the (n-1, n-1)
+                # system matrix indices forming their names:
+                a_m1_0 = dx[-2]  # lower left corner value: A[-1, 0]
+                a_m1_m2 = dx[-1]
+                a_m1_m1 = 2 * (dx[-1] + dx[-2])
+                a_m2_m1 = dx[-2]
+                a_0_m1 = dx[0]
+
+                b[0] = 3 * (dxr[0] * slope[-1] + dxr[-1] * slope[0])
+                b[-1] = 3 * (dxr[-1] * slope[-2] + dxr[-2] * slope[-1])
+
+                Ac = A[:, :-1]
+                b1 = b[:-1]
+                b2 = np.zeros_like(b1)
+                b2[0] = -a_0_m1
+                b2[-1] = -a_m2_m1
+
+                # s1 and s2 are the solutions of (n-2, n-2) system
+                s1 = solve_banded((1, 1), Ac, b1, overwrite_ab=False,
+                                  overwrite_b=False, check_finite=False)
+
+                s2 = solve_banded((1, 1), Ac, b2, overwrite_ab=False,
+                                  overwrite_b=False, check_finite=False)
```

Comment:
(Non-binding comment) these can be packed into a single call to `solve_banded` with r.h.s. with shape (N, 2).